### PR TITLE
Refactor and implement `Restart-Computer` for `Un*x` and macOS

### PR DESCRIPTION
**Pull request by brendandburns**
# PR Summary
Implement the `Restart-Computer` command for Un*x and MacOS

## PR Context
Compliment to the recent implementation of `Stop-Computer` (#11151)

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
- **User-facing changes**
    - [X] Not Applicable
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs [hunk 1]
```diff
@@ -10,22 +10,70 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region Restart-Computer
+
+    /// <summary>
+    /// Cmdlet to restart computer.
+    /// </summary>
+    [Cmdlet(VerbsLifecycle.Restart, "Computer", SupportsShouldProcess = true,
+        HelpUri = "https://go.microsoft.com/fwlink/?LinkID=135263", RemotingCapability = RemotingCapability.SupportedByCommand)]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]: Please use new value for `LinkID=135263`

**Comment by iSazonov** [comment]
@brendandburns Please don't rebase until maintainers ask you. Rebasing complicates reviewing.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs [hunk 2]
```diff
@@ -10,22 +10,70 @@
...
-        private Process _process = null;
+        #region "Overrides"
 
-        #endregion
+        /// <summary>
+        /// BeginProcessing.
+        /// </summary>
+        protected override void BeginProcessing()
+        {
+            var args = "";
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]: Perhaps initialize `args = "-P now"` to optimize for Linux and just check for macOS?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs [hunk 3]
```diff
@@ -10,22 +10,70 @@
...
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                args = "-P now";
+            }
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                args = "now";
+            }
+            RunCommand("/sbin/shutdown", args);
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]: The style rules we have expect a newline after closing brace
```suggestion

            RunCommand("/sbin/shutdown", args);
```

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
The Windows Computer cmdlets already use a test hook for testing, should use the same thing here.  See https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs#L2130

**Comment by iSazonov** [comment]
We merged #11151 without new tests.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs (hunk 1 above)
Comment (brendandburns) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]: Done.

**Comment by brendandburns** [comment]
I will address comments and add tests for both commands.

Thanks

**Comment by brendandburns** [comment]
@iSazonov is there a preferred style of addressing comments for the PowerShell repo? Should I just add additional commits instead of squashing commits?

(I'm new here :)

Thanks

**Comment by iSazonov** [comment]
> Should I just add additional commits instead of squashing commits?

Yes, please add new commits. Some commits or one for all comments.

**Comment by SteveL-MSFT** [comment]
@brendandburns unless there is a need to preserve history, we always squash on merge.  For the sake of the reviewers, it's always been easier to add new commits so we can see what exactly changed.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1 [hunk 4]
```diff
@@ -10,7 +10,7 @@ $DefaultResultValue = 0
 try
 {
     # set up for testing
-    $PSDefaultParameterValues["it:skip"] = ! $IsWindows
+    # $PSDefaultParameterValues["it:skip"] = ! $IsWindows
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1]: Please remove the commented code.

Comment (brendandburns) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1]: Done.

**Comment by brendandburns** [comment]
@iSazonov @SteveL-MSFT 

Sorry it took so long for me to get back to this. I've added tests for both `Restart-Computer` and `Stop-Computer` on Unix.

Please take another look.

Thanks!

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs (hunk 3 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]: @brendandburns Please address the comment.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs [hunk 5]
```diff
@@ -6,26 +6,97 @@
...
+            }
 
-        // TODO: Support remote computers?
+            RunCommand("/sbin/shutdown", args);
+        }
+        #endregion "Overrides"
+    }
+
+    /// <summary>
+    /// A base class for cmdlets that can run shell commands
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]:
```suggestion
    /// A base class for cmdlets that can run shell commands.
```

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs [hunk 6]
```diff
@@ -67,21 +129,15 @@ protected override void StopProcessing()
...
+        #endregion Overrides
 
-        #endregion "Overrides"
+        #region Methods
 
-        #region "Internals"
-
-        private void doShutdown() {
+        /// <summary>
+        /// Run a command
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]:
```suggestion
        /// Run a command.
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1 [hunk 7]
```diff
@@ -30,12 +29,18 @@ try
         }
 
         It "Should support -computer parameter" {
+            if ( ! $IsWindows ) {
+                return
+            }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1]: Please use `It "Should support -computer parameter" -Skip:(!$IsWindows) {`

File: test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1 [hunk 8]
```diff
@@ -30,12 +29,18 @@ try
...
             Set-TesthookResult -testhookName $restartTesthookResultName -value $defaultResultValue
             $computerNames = "localhost","${env:COMPUTERNAME}"
             Restart-Computer -Computer $computerNames -ErrorAction Stop | Should -BeNullOrEmpty
         }
 
         It "Should support WsmanAuthentication types" {
+            if ( ! $IsWindows ) {
+                return
+            }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1]: The same.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1 [hunk 9]
```diff
@@ -46,6 +51,9 @@ try
         # set operation. Internally, we want to suppress the progress, so
         # that is also wrapped in try/finally
         It "Should wait for a remote system" {
+            if ( ! $IsWindows ) {
+                return
+            }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1]: The same.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs (hunk 3 above)
Comment (brendandburns) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]: Done.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1 (hunk 8 above)
Comment (brendandburns) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1]: done for all.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1 (hunk 7 above)
Comment (brendandburns) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1]: done for all.

**Comment by brendandburns** [comment]
@iSazonov comments addressed, please re-check.

Thanks!

**Comment by iSazonov** [comment]
@brendandburns Did you address all @SteveL-MSFT 's comments?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs (hunk 2 above)
Comment (brendandburns) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs]: done.

**Comment by brendandburns** [comment]
yes, I verified in the code that I had addressed both of the comments from @SteveL-MSFT 

Thanks

**Comment by iSazonov** [comment]
@SteveL-MSFT Please update your review.

**Comment by TravisEz13** [comment]
@SteveL-MSFT Please update your review.

**Comment by TravisEz13** [comment]
@SteveL-MSFT  Please update your review

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by TravisEz13** [comment]
The merge is bad.  I'll fix it.

**Comment by TravisEz13** [comment]
Rebased to resolve conflicts and changed #region format to match changes from conflict

**Comment by TravisEz13** [comment]
@PoshChan Please remind me in 1 hour

**Comment by PoshChan** [comment]
@TravisEz13, this is the reminder you requested 1 hour ago

**Comment by iSazonov** [comment]
@brendandburns Thanks for your contribution!

**Pull request merged**